### PR TITLE
[client-rust] Add upload/download functionality

### LINKED ISSUE #4422: Add support for uploading and downloading to the Rust client

### PR DESCRIPTION
**Pull request by djmitche**
Github Bug/Issue: Fixes #4422

This is roughly divided into commits.  It entails a full refactoring of the Rust client into multiple crates, so that users can opt-in to the download/upload functionality and the dependencies it entails.

There's an `unsafe` block in here that I'm both not sure about and not sure how to fix.  Advice on who to ask for review there would be good!

### CODE REVIEW

**Comment by djmitche** [comment]
After some discussion in #rust, I'm going to refactor a it to avoid the `unsafe` bit by doing some cloning, first trying to use an `Arc<Vec<u8>>` to clone once, and falling back to cloning on every `get_reader` call.

I also forgot to add credentials to the test suite!

**Comment by djmitche** [comment]
`Arc<Vec<u8>>` doesn't implement the `AsRef<[u8]>` that `Cursor` requires, so I elected to just clone.

**Comment by djmitche** [comment]
OK, this is now failing only because there's no Object services in the community deployment.  I'll get that fixed up on upgrade.

**Comment by djmitche** [comment]
We should probably have a way for the caller to get the content-type, as well.

**Comment by djmitche** [comment]
`service-auth` failure is #3665 -- the client-rust test passed :)

**Review by imbstack — COMMENTED** [review]
This is going to take me a while to review if I'm going to do a good job on it. Is that ok?

File: changelog/IuNWJ07pRtiW4b1faDHL8w.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+audience: general
```
[the remark this answers is not in the document]
Comment (djmitche) [reply changelog/IuNWJ07pRtiW4b1faDHL8w.md:1]: Oh, probably a merge error -- I think I cherry-picked that out into another PR.  I'll drop the commit.

**Comment by djmitche** [comment]
It's fine to take a while.  This should probably land at about the same time as the other client changes, including go, and that's not ready for review yet.

**Review by imbstack — APPROVED** [review]
I think this seems reasonable but tbh my mind is not able to get into this enough to really give it a good review now. Maybe want to block landing on @leplatrem

**Comment by djmitche** [comment]
@leplatrem what do you think?  The consequences of a bug here are also pretty minor at the moment since this library is brand new.

**Comment by djmitche** [comment]
This is blocking #4608, so I'm going to merge it.  Please feel free to review anyway -- we can address any issues or notes in further PRs.

**Pull request merged**